README.md
# Vdebug

[![Build Status](https://travis-ci.org/joonty/vdebug.png?branch=master)](https://travis-ci.org/joonty/vdebug)

## Introduction

Vdebug is a new, fast, powerful debugger client for Vim. It's multi-language,
and has been tested with PHP, Python, Ruby, Perl, Tcl and NodeJS. It interfaces with 
**any** debugger that faithfully uses the DBGP protocol, such as Xdebug for PHP.  
There are step-by-step instructions for setting up debugging with all of the aforementioned 
languages in the Vim help file that comes with Vdebug. 

It builds on the experience gained through the legacy of the Xdebug Vim script 
originally created by Seung Woo Shin and extended by so many others, but it's a
total rebuild to allow for a nicer interface and support of new features.

It's written in Python, and has an object-oriented interface that is easy to extend 
and can even be used from the command-line. It even has unit tests covering
some of the more critical parts of the code.

## Getting help

If you're having trouble with Vdebug in any way, here are the steps you can take to get help (in the right order):

  1. [Check the issues][3] to see whether it's already come up.
  2. Visit the **#vdebug** irc channel on freenode, someone is normally there.
  3. [Open a new issue.][4]

## News

Follow me on Twitter, [@joonty](http://twitter.com/joonty), for Vdebug updates and release news.

* **26/08/2015:** Version 1.5.0 released. This has a whole series of fixes and even a couple of new features. See below for more information.
* **15/05/2014:** Version 1.4.2 released. This fixes major problems with the latest version of Vim, which has stricter rules around function naming. See [this pull request for more details][5]
* **12/04/2014:** My son was born! Hence no contact or work on Vdebug for the following month.

## Recent version (version 1.5.0)

### Bug fixes:
 * Fix for remote and local file paths with a windows/unix setup (thanks @brothert, #130)
 * Sort path maps by remote path length (thanks @ragol and @adambiggs, #153, #201)
 * Default "stop" behaviour (default <F6>) actually stops rather than detaches (issue #166)
 * Various documentation fixes (issue #172, issue #223, thanks @loonies for #210)
 * Add documentation to show that komodo-debug doesn't work with nodejs > 0.10 (#202, #223)
 * Don't override previously defined breakpoint highlights (#208)

### Features:
 * Add support for Python 2.4 (thanks @mbarbon, #189)
 * Add Vagrantfile and bootstrap for easier local development
 * Add new VdebugTrace command for tracking individual variables and expressions (thanks @escher9, #178)
 * Show ellipses for truncated arrays in watch window (#177)
 * Bind to all available interfaces by default, instead of just localhost (#209, thanks @zolem)

### Misc:
 * Changed HISTORY to CHANGELOG

## How to use

First of all, scoot down to the quick guide below.

There is *extensive* help provided in the form of a Vim help file. This goes
through absolutely everything, from installation to configuration, setting up
debuggers for various languages, explanation of the interface, options, remote
server debugging and more.

To get this help, type:

```
:help Vdebug
```

after installing the plugin. If this doesn't work, you will need to generate the help tags.

### Generating help tags

Some package managers, such as Vundle, will automatically generate help tags. Others, such as Pathogen, won't by default, so you have to do it yourself. Also, if you have installed Vdebug manually then you will also have to do this step.

For pathogen users, you can add this to your vimrc file to automatically generate helptags for your plugins:

```vim
call pathogen#helptags()
```

To manually generate the tags, run this in vim:

```vim
:helptags /path/to/vdebug/doc
```

where the path supplied is vdebug's doc directory. This should enable vdebug's help to be accessed.

## Installation

**Requirements**:

  * Vim compiled with Python 2.6+ support, tabs and signs (for Debian/Ubuntu this is provided in the vim-nox package)
  * A programming language that has a DBGP debugger, e.g. PHP, Python, Ruby,
    Perl, NodeJS, Tcl...

### Classic

Clone or download a tarball of the plugin and move its content in your
`~/.vim/` directory.

Your `~/.vim/plugin/` directory should now contain vdebug.vim and a directory
called "python".

### Using git and Pathogen

Clone this repository in your `~/.vim/bundle` directory

### Using vundle

Add this to your `~/.vimrc` file:

```vim
Bundle 'joonty/vdebug.git'
```

Then, from the command line, run:

```bash
vim +BundleInstall +qall
```

## Quick guide

Set up any DBGP protocol debugger, e.g. Xdebug. (See :help VdebugSetUp). Start Vdebug with `<F5>`, which will make it wait for an incoming connection. Run the script you want to debug, with the debugging engine enabled. A new tab will open with the debugging interface.

Once in debugging mode, the following default mappings are available:

 * `<F5>`: start/run (to next breakpoint/end of script)
 * `<F2>`: step over
 * `<F3>`: step into
 * `<F4>`: step out
 * `<F6>`: stop debugging (kills script)
 * `<F7>`: detach script from debugger
 * `<F9>`: run to cursor
 * `<F10>`: toggle line breakpoint
 * `<F11>`: show context variables (e.g. after "eval")
 * `<F12>`: evaluate variable under cursor
 * `:Breakpoint <type> <args>`: set a breakpoint of any type (see :help
    VdebugBreakpoints)
 * `:VdebugEval <code>`: evaluate some code and display the result
 * `<Leader>e`: evaluate the expression under visual highlight and display the result

To stop debugging, press `<F6>`. Press it again to close the debugger interface.

If you can't get a connection, then chances are you need to spend a bit of time setting up your environment. Type `:help Vdebug` for more information.

## Debugging

If you have a problem, and would like to see what's going on under the hood or raise an issue, it's best to create a log file. You can do this by setting these options before you start debugging:

```vim
:VdebugOpt debug_file ~/vdebug.log
:VdebugOpt debug_file_level 2
```

Then start debugging, and you can follow what's added to the log file as you go. It shows the communication between the debugging engine and Vdebug.

If you're creating an issue then it's probably best to upload a log as a Gist, as it can be pretty large.

## Contributing

I gladly accept contributions to the code. Just fork the repository, make your changes and open a pull request with detail about your changes. There are a couple of conditions:

 * The tests must pass (see below)
 * Your commit messages should follow the [rules outlined here][2]

## Tests

 * The tests use `unittest2` and `mock`, so make sure they're installed

```
pip install unittest2
pip install mock
```

* To run the tests, run `python vdebugtests.py` in the top directory of the plugin

## Licence

This plugin is released under the [MIT License][1].

[1]: https://raw.github.com/joonty/vdebug/master/LICENCE
[2]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[3]: https://github.com/joonty/vdebug/issues/
[4]: https://github.com/joonty/vdebug/issues/new
[5]: https://github.com/joonty/vdebug/issues/158


HISTORY
=== Vdebug 1.5.0 / 2015-08-28
Bug fixes:
 * Fix for remote and local file paths with a windows/unix setup (thanks @brothert, #130)

Features:
 * Add new VdebugTrace command for tracking individual variables and expressions (thanks @escher9, #178)

=== Vdebug 1.4.2 / 2014-05-15

Bug fixes:
 * Allow overriding of options dictionary after sourcing vdebug (issues #112, #113)
 * Show local file path in stack window if using path mapping (issue #118)
 * Change function names for stricter rules introduced in Vim v.7.4.260 (thanks @BlackEagle, #158 and #155)

Documentation:
 * Various fixes
 * Mention vim-nox package for Debian (thanks @henningda, #138)
 * Add Python installation for Windows information (issue #135)

Features:
 * Add completion to :Breakpoint command (thanks @tommcdo, #124)
 * Add VdebugStart command (thanks @blueyed, #132)
 * Sleep for 100ms in wait loop, don't consume as much CPU (thanks @blueyed, #133)

=== Vdebug 1.4.1 / 2013-06-09

Bug fixes:
 * Fix folds being forgotten after debugging (issue #56)
 * Don't overwrite features array if it exists (issue #59)
 * Mapping of single modes (thanks @xbot, #66)
 * Stop empty buffers from being created (issue #70)
 * Fix ElementTree deprecation warning (thanks @InFog, #89)
 * Fix minheight error for windows (thanks @zhaocai, #73, also #84)
 * Fix filename handling on windows (thanks @Chronial, #74)
 * Use :mkexrc for restoring keymapping (thanks @Chronial and @qstrahl, #75)
 * Fix window destruction (thanks @Chronial, #76)
 * Inspect PHP constants with cursor eval (issue #80)
 * Don't cut final character when inspecting variables at end of line (issue #85)

Documentation:
 * Test suite requirements (thanks @michaeltwofish, #61)
 * Add helptag instructions to README (issue #67)
 * Add 'eval_visual' to documentation (issue #88)

=== Vdebug 1.4.0 / 2013-03-06

Features:
 * Set debugger features with g:vdebug_features dictionary (issue #55)
 * Allow use of '~' in log files (issue #53)

Bug fixes:
 * Print error message if trying to start debugger with unsaved buffers (issue #57)
 * Don't allow visual eval outside of debugging session (issue #52)
 * Show updated key mappings in status window (issue #51)
 * Fix problem with stuck breakpoints after debugging (issue #48, #43)
 * Swap back to original tab after debugging session closes (issue #23)

=== Vdebug 1.3.2 / 2013-01-28

Features:
 * Don't allow breakpoints to be set on empty lines (issue #27)
 * Add :VdebugOpt command for easy setting and getting of options (issue #38)

Bug fixes:
 * Small fixes for documentation (thanks: https://github.com/grota, #31)
 * Parse continuous mode option as an integer (thanks: https://github.com/grota, #32)
 * Stop Xdebug hanging when using the detach command (issue #33)
 * Don't use recursive Vim mapping (issue #34)
 * Track movement of line breaks with Vim signs (issue #35)
 * Add "default" keyword to highlights, to allow custom colours (issue #36)
 * Fix check for multibyte support (issue #37)
 * Fix deprecation warning in Python 2.6, catching string exceptions (issue #39)
 * Stop source code panel opening in incorrect position (issue #40)

=== Vdebug 1.3.1 / 2012-12-21

Bug fixes:
 * Filepath problems with Windows version

Other:
 * Packaged tests as an easy-to-run suite
 * Set up on Travis CI

=== Vdebug 1.3.0 / 2012-12-19

Features:
 * Added minor version number to versioning, for bug fixes
 * Support and compatibility fixes for debugging Tcl, plus help file updates
 * Non-multibyte Vim installations now use ASCII watch window marker characters as a fallback (thanks: https://github.com/satiani, #21)
 * Watch window marker characters are fully customizable
 * Added breakpoint list function to Api class
 * Satiani added multiple file path mappings, for remote debugging (thanks: https://github.com/satiani, #20)
 * Vdebug re-maps previous key mappings after shutting down (issue #16)
 * Added 'watch_window_style' option that allows for a compact or (default) expanded watch window (thanks: https://github.com/georgjaehnig, #13)
 * Simple continuous mode added, where vdebug starts listening after a connection closes, to allow for debugging to start again immediately (issue #19)

Bug fixes:
 * Fix for watch window characters on Windows versions of Vim (issue #29)
 * Fix for watch window variable expansion on Windows versions of Vim (issue #28)
 * Stopped creating empty buffers every time Vdebug started (issue #25)
 * Fixed bug with file URI quoting
 * Fixed escaping of characters when calling eval (issue #15)
 

=== Vdebug 1.2 / 2012-10-05

 * Support for nodejs debugging using the komodo-debug package in NPM
 * Keyboard interrupt while Vim is waiting for a connection (thanks: http://github.com/artnez, #6)
 * Fixed overwriting dir function with badly named variable (thanks: http://github.com/csomme, #8)

plugin/python/dbgp.py
import xml.etree.ElementTree as ET
import socket
import log
import base64

""" Response objects for the DBGP module."""

class Response:
    """Contains response data from a command made to the debugger."""
    ns = '{urn:debugger_protocol_v1}'

    def __init__(self,response,cmd,cmd_args,api):
        self.response = response
        self.cmd = cmd
        self.cmd_args = cmd_args
        self.xml = None
        self.api = api
        if "<error" in self.response:
            self.__parse_error()

    def __parse_error(self):
        """Parse an error message which has been returned
        in the response, then raise it as a DBGPError."""
        xml = self.as_xml()
        err_el = xml.find('%serror' % self.ns)
        code = err_el.get("code")
        if code is None:
            raise ResponseError(
                    "Missing error code in response",
                    self.response)
        msg_el = err_el.find('%smessage' % self.ns)
        if msg_el is None:
            raise ResponseError(
                    "Missing error message in response",
                    self.response)
        raise DBGPError(msg_el.text,code)

    def get_cmd(self):
        """Get the command that created this response."""
        return self.cmd

    def get_cmd_args(self):
        """Get the arguments to the command."""
        return self.cmd_args

    def as_string(self):
        """Return the full response as a string.
        
        There is a __str__ method, which will render the
        whole object as a string and should be used for
        displaying.
        """
        return self.response

    def as_xml(self):
        """Get the response as element tree XML.

        Returns an xml.etree.ElementTree.Element object.
        """
        if self.xml == None:
            self.xml = ET.fromstring(self.response)
        return self.xml

    def __str__(self):
        return self.as_string()

class ContextNamesResponse(Response):
    def names(self):
        names = {}
        for c in self.as_xml().getchildren():
            names[int(c.get('id'))] = c.get('name')
        return names

class StatusResponse(Response):
    """Response object returned by the status command."""

    def __str__(self):
        return self.as_xml().get('status')

class StackGetResponse(Response):
    """Response object used by the stack_get command."""

    def get_stack(self):
        return self.as_xml().getchildren()

class ContextGetResponse(Response):
    """Response object used by the context_get command.
    
    The property nodes are converted into ContextProperty
    objects, which are much easier to use."""

    def __init__(self,response,cmd,cmd_args,api):
        Response.__init__(self,response,cmd,cmd_args,api)
        self.properties = []

    def get_context(self):
        for c in self.as_xml().getchildren():
            self.create_properties(ContextProperty(c))

        return self.properties

    def create_properties(self,property):
        self.properties.append(property)
        for p in property.children:
            self.create_properties(p)

class EvalResponse(ContextGetResponse):
    """Response object returned by the eval command."""

    def get_context(self):
        code = self.get_code()
        for c in self.as_xml().getchildren():
            self.create_properties(EvalProperty(c,code,self.api.language))

        return self.properties

    def get_code(self):
        cmd = self.get_cmd_args()
        parts = cmd.split('-- ')
        return base64.decodestring(parts[1])


class BreakpointSetResponse(Response):
    """Response object returned by the breakpoint_set command."""

    def get_id(self):
        return int(self.as_xml().get('id'))

    def __str__(self):
        return self.as_xml().get('id')

class FeatureGetResponse(Response):
    """Response object specifically for the feature_get command."""

    def is_supported(self):
        """Whether the feature is supported or not."""
        xml = self.as_xml()
        return int(xml.get('supported'))

    def __str__(self):
        if self.is_supported():
            xml = self.as_xml()
            return xml.text
        else:
            return "* Feature not supported *"

class Api:
    """Api for eBGP commands.

    Uses a Connection object to read and write with the debugger,
    and builds commands and returns the results.
    """

    conn = None
    transID = 0

    def __init__(self,connection):
        """Create a new Api using a Connection object.

        The Connection object specifies the debugger connection,
        and the Protocol provides a OO api to interacting
        with it.

        connection -- The Connection object to use
        """
        self.language = None
        self.protocol = None
        self.idekey = None
        self.startfile = None
        self.conn = connection
        if self.conn.isconnected() == 0:
            self.conn.open()
        self.__parse_init_msg(self.conn.recv_msg())
        
    def __parse_init_msg(self,msg):
        """Parse the init message from the debugger"""
        xml = ET.fromstring(msg)
        self.language = xml.get("language")
        if self.language is None:
            raise ResponseError(
                "Invalid XML response from debugger",
                msg)

        self.idekey = xml.get("idekey")
        self.version = xml.get("api_version")
        self.startfile = xml.get("fileuri")

    def send_cmd(self,cmd,args = '',
            res_cls = Response):
        """Send a command to the debugger.

        This method automatically adds a unique transaction
        ID to the command which is required by the debugger.

        Returns a Response object, which contains the
        response message and command.

        cmd -- the command name, e.g. 'status'
        args -- arguments for the command, which is optional 
                for certain commands (default '')
        """
        args = args.strip()
        send = cmd.strip()
        self.transID += 1
        send += ' -i '+ str(self.transID)
        if len(args) > 0:
            send += ' ' + args
        log.Log("Command: "+send,\
                log.Logger.DEBUG)
        self.conn.send_msg(send)
        msg = self.conn.recv_msg()
        log.Log("Response: "+msg,\
                log.Logger.DEBUG)
        return res_cls(msg,cmd,args,self)

    def status(self):
        """Get the debugger status.
        
        Returns a Response object.
        """
        return self.send_cmd('status','',StatusResponse)

    def feature_get(self,name):
        """Get the value of a feature from the debugger.

        See the DBGP documentation for a list of features.
        
        Returns a FeatureGetResponse object.
        
        name -- name of the feature, e.g. encoding
        """
        return self.send_cmd(
                'feature_get',
                '-n '+str(name),
                FeatureGetResponse)

    def feature_set(self,name,value):
        """Set the value of a debugger feature.

        See the DBGP documentation for a list of features.
        
        Returns a Response object.
        
        name -- name of the feature, e.g. encoding
        value -- new value for the feature
        """
        return self.send_cmd(
                'feature_set',
                '-n ' + str(name) + ' -v ' + str(value))

    def run(self):
        """Tell the debugger to start or resume
        execution."""
        return self.send_cmd('run','',StatusResponse)

    def eval(self,code):
        """Tell the debugger to start or resume
        execution."""
        code_enc = base64.encodestring(code)
        return self.send_cmd('eval','-- %s' % code_enc,EvalResponse)

    def step_into(self):
        """Tell the debugger to step to the next
        statement.
        
        If there's a function call, the debugger engine
        will break on the first statement in the function.
        """
        return self.send_cmd('step_into','',StatusResponse)

    def step_over(self):
        """Tell the debugger to step to the next
        statement.
        
        If there's a function call, the debugger engine
        will stop at the next statement after the function call.
        """
        return self.send_cmd('step_over','',StatusResponse)

    def step_out(self):
        """Tell the debugger to step out of the statement.
        
        The debugger will step out of the current scope.
        """
        return self.send_cmd('step_out','',StatusResponse)

    def stop(self):
        """Tell the debugger to stop execution.

        The script is terminated immediately."""
        return self.send_cmd('stop','',StatusResponse)

    def stack_get(self):
        """Get the stack information.
        """
        return self.send_cmd('stack_get','',StackGetResponse)

    def context_get(self,context = 0):
        """Get the context variables.
        """
        return self.send_cmd('context_get',\
                '-c %i' % int(context),\
                ContextGetResponse)

    def context_names(self):
        """Get the context types.
        """
        return self.send_cmd('context_names','',ContextNamesResponse)

    def property_get(self,name):
        """Get a property.
        """
        return self.send_cmd('property_get','-n '+name,ContextGetResponse)

    def detach(self):
        """Tell the debugger to detach itself from this
        client.

        The script is not terminated, but runs as normal
        from this point."""
        return self.send_cmd('detach','',StatusResponse)

    def breakpoint_set(self,cmd_args):
        """Set a breakpoint.

        The breakpoint type is defined by the arguments, see the
        Breakpoint class for more detail."""
        return self.send_cmd('breakpoint_set',cmd_args,\
                BreakpointSetResponse)

    def breakpoint_remove(self,id):
        """Remove a breakpoint by ID.

        The ID is that returned in the response from breakpoint_set."""
        return self.send_cmd('breakpoint_remove','-d %i' % id,Response)

"""Connection module for managing a socket connection
between this client and the debugger."""

class Connection:
    """DBGP connection class, for managing the connection to the debugger.

    The host, port and socket timeout are configurable on object construction.
    """

    sock = None
    address = None
    isconned = 0

    def __init__(self, host = '', port = 9000, timeout = 30):
        """Create a new Connection.

        The connection is not established until open() is called.

        host -- host name where debugger is running (default '')
        port -- port number which debugger is listening on (default 9000)
        timeout -- time in seconds to wait for a debugger connection before giving up (default 30)
        """
        self.port = port
        self.host = host
        self.timeout = timeout

    def __del__(self):
        """Make sure the connection is closed."""
        self.close()

    def isconnected(self):
        """Whether the connection has been established."""
        return self.isconned

    def open(self):
        """Listen for a connection from the debugger.

        The socket is blocking, and it will wait for the length of
        time given by the timeout (default is 30 seconds).
        """
        print 'Waiting for a connection (this message will self-destruct in ',self.timeout,' seconds...)'
        serv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            serv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            serv.settimeout(self.timeout)
            serv.bind((self.host, self.port))
            serv.listen(5)
            (self.sock, self.address) = serv.accept()
            self.sock.settimeout(None)
        except socket.timeout:
            serv.close()
            raise TimeoutError,"Timeout waiting for connection"

        self.isconned = 1
        serv.close()

    def close(self):
        """Close the connection."""
        if self.sock != None:
            self.sock.close()
            self.sock = None
        self.isconned = 0

    def __recv_length(self):
        """Get the length of the proceeding message."""
        length = ''
        while 1:
            c = self.sock.recv(1)
            if c == '':
                self.close()
                raise EOFError, 'Socket Closed'
            if c == '\0':
                return int(length)
            if c.isdigit():
                length = length + c

    def __recv_null(self):
        """Receive a null byte."""
        while 1:
            c = self.sock.recv(1)
            if c == '':
                self.close()
                raise EOFError, 'Socket Closed'
            if c == '\0':
                return

    def __recv_body(self, to_recv):
        """Receive a message of a given length.

        to_recv -- length of the message to receive
        """
        body = ''
        while to_recv > 0:
            buf = self.sock.recv(to_recv)
            if buf == '':
                self.close()
                raise EOFError, 'Socket Closed'
            to_recv -= len(buf)
            body = body + buf
        return body

    def recv_msg(self):
        """Receive a message from the debugger.
        
        Returns a string, which is expected to be XML.
        """
        length = self.__recv_length()
        body     = self.__recv_body(length)
        self.__recv_null()
        return body

    def send_msg(self, cmd):
        """Send a message to the debugger.

        cmd -- command to send
        """
        self.sock.send(cmd + '\0')

class ContextProperty:

    ns = '{urn:debugger_protocol_v1}'

    def __init__(self,node,parent = None,depth = 0):
        self.parent = parent
        self.__determine_type(node)
        self._determine_displayname(node)
        self.encoding = node.get('encoding')
        self.depth = depth
        self.size = node.get('size')
        self.value = ""
        self.is_last_child = False

        self._determine_children(node)
        self.__determine_value(node)
        self.__init_children(node)

    def __determine_value(self,node):
        if self.has_children:
            self.value = ""
            return

        self.value = self.__get_enc_node_text(node,'value')
        if self.value is None:
            if self.encoding == 'base64':
                if node.text is None:
                    self.value = ""
                else:
                    self.value = base64.decodestring(node.text)
            elif not self.is_uninitialized() \
                    and not self.has_children:
                self.value = node.text

        if self.value is None:
            self.value = ""

        self.num_crs = self.value.count('\n')
        if self.type.lower() in ("string","str"):
            self.value = '`%s`' % self.value.replace('`','\\`')

    def __determine_type(self,node):
        type = node.get('classname')
        if type is None:
            type = node.get('type')
        self.type = type

    def _determine_displayname(self,node):
        display_name = node.get('fullname')
        if display_name == None:
            display_name = self.__get_enc_node_text(node,'fullname',"")
        if display_name == '::':
            display_name = self.type
        self.display_name = display_name

    def __get_enc_node_text(self,node,name,default =
            None):
        n = node.find('%s%s' %(self.ns, name))
        if n is not None and n.text is not None:
            if n.get('encoding') == 'base64':
                val = base64.decodestring(n.text)
            else:
                val = n.text
        else:
            val = None
        if val is None:
            return default
        else:
            return val

    def _determine_children(self,node):
        children = node.get('numchildren')
        if children is None:
            children = node.get('children')
        if children is None:
            children = 0
        else:
            children = int(children)
        self.num_declared_children = children
        self.has_children = True if children > 0 else False
        self.children = []

    def __init_children(self,node):
        if self.has_children:
            idx = 0
            tagname = '%sproperty' % self.ns
            children = node.getchildren()
            if children is not None:
                for c in children:
                    if c.tag == tagname:
                        idx += 1
                        p = self._create_child(c,self,self.depth+1)
                        self.children.append(p)
                        if idx == self.num_declared_children:
                            p.mark_as_last_child()

    def _create_child(self,node,parent,depth):
        return ContextProperty(node,parent,depth)

    def mark_as_last_child(self):
        self.is_last_child = True

    def is_uninitialized(self):
        if self.type == 'uninitialized':
            return True
        else:
            return False

    def child_count(self):
        return len(self.children)

    def type_and_size(self):
        size = None
        if self.has_children:
            size = self.num_declared_children
        elif self.size is not None:
            size = self.size

        if size is None:
            return self.type
        else:
            return "%s [%s]" %(self.type,size)

class EvalProperty(ContextProperty):
    def __init__(self,node,code,language,parent=None,depth=0):
        self.code = code
        self.language = language.lower()
        if parent is None:
            self.is_parent = True
        else:
            self.is_parent = False
        ContextProperty.__init__(self,node,parent,depth)

    def _create_child(self,node,parent,depth):
        return EvalProperty(node,self.code,self.language,parent,depth)

    def _determine_displayname(self,node):
        if self.is_parent:
            self.display_name = "(%s)" % self.code
        else:
            if self.language == 'php' or \
                    self.language == 'perl':
                if self.parent.type == 'array':
                    self.display_name = self.parent.display_name + \
                        "['%s']" % node.get('name')
                else:
                    self.display_name = self.parent.display_name + \
                        "->"+node.get('name')
            else:
                self.display_name = self.parent.display_name + \
                    "." + node.get('name')


""" Errors/Exceptions """
class TimeoutError(Exception):
    pass


class DBGPError(Exception):
    """Raised when the debugger returns an error message."""
    pass

class ResponseError(Exception):
    """An error caused by an unexpected response from the
    debugger (e.g. invalid format XML)."""
    pass

error_codes = { \
    # 000 Command parsing errors
    0   : """no error""",\
    1   : """parse error in command""",\
    2   : """duplicate arguments in command""", \
    3   : """invalid options (ie, missing a required option)""",\
    4   : """Unimplemented command""",\
    5   : """Command not available (Is used for async commands. For instance if the engine is in state "run" than only "break" and "status" are available). """,\
    # 100 : File related errors
    100 : """can not open file (as a reply to a "source" command if the requested source file can't be opened)""",\
    101 : """stream redirect failed """,\
    # 200 Breakpoint, or code flow errors
    200 : """breakpoint could not be set (for some reason the breakpoint could not be set due to problems registering it)""",\
    201 : """breakpoint type not supported (for example I don't support 'watch' yet and thus return this error)""",\
    202 : """invalid breakpoint (the IDE tried to set a breakpoint on a line that does not exist in the file (ie "line 0" or lines past the end of the file)""",\
    203 : """no code on breakpoint line (the IDE tried to set a breakpoint on a line which does not have any executable code. The debugger engine is NOT required to """     + \
          """return this type if it is impossible to determine if there is code on a given location. (For example, in the PHP debugger backend this will only be """         + \
          """returned in some special cases where the current scope falls into the scope of the breakpoint to be set)).""",\
    204 : """Invalid breakpoint state (using an unsupported breakpoint state was attempted)""",\
    205 : """No such breakpoint (used in breakpoint_get etc. to show that there is no breakpoint with the given ID)""",\
    206 : """Error evaluating code (use from eval() (or perhaps property_get for a full name get))""",\
    207 : """Invalid expression (the expression used for a non-eval() was invalid) """,\
    # 300 Data errors
    300 : """Can not get property (when the requested property to get did not exist, this is NOT used for an existing but uninitialized property, which just gets the """    + \
          """type "uninitialised" (See: PreferredTypeNames)).""",\
    301 : """Stack depth invalid (the -d stack depth parameter did not exist (ie, there were less stack elements than the number requested) or the parameter was < 0)""",\
    302 : """Context invalid (an non existing context was requested) """,\
    # 900 Protocol errors
    900 : """Encoding not supported""",\
    998 : """An internal exception in the debugger occurred""",\
    999 : """Unknown error """\
}

